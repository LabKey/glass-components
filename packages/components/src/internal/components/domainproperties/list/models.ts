/*
 * Copyright (c) 2019 LabKey Corporation
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
import { Record } from 'immutable';

import { DomainDesign, DomainField } from '../models';
import { DOMAIN_FIELD_PRIMARY_KEY_LOCKED } from '../constants';

import { PRIVATE_PICKLIST_CATEGORY, PUBLIC_PICKLIST_CATEGORY } from '../../picklist/constants';

import { INT_LIST, PICKLIST, VAR_LIST } from './constants';

export interface EachItemSettings {
    eachItemBodySetting: number;
    eachItemBodyTemplate: string;
    eachItemTitleTemplate: string;
}

export interface EntireListSettings {
    entireListBodySetting: number;
    entireListBodyTemplate: string;
    entireListIndexSetting: number;
    entireListTitleTemplate: string;
}

export interface AdvancedSettingsForm extends EachItemSettings, EntireListSettings {
    discussionSetting: number;
    eachItemIndex: boolean;
    entireListIndex: boolean;
    fileAttachmentIndex: boolean;
    multiFolder: boolean;
    titleColumn: string;
}

export class ListModel extends Record({
    exception: undefined,
    domain: undefined,
    entityId: undefined,
    createdBy: undefined,
    created: undefined,
    modifiedBy: undefined,
    modified: undefined,
    containerId: undefined,
    name: undefined,
    domainKindName: undefined,
    description: undefined,
    lastIndexed: undefined,
    keyName: undefined,
    titleColumn: undefined,
    domainId: undefined,
    keyType: undefined,
    discussionSetting: undefined,
    allowDelete: undefined,
    allowUpload: undefined,
    allowExport: undefined,
    entireListIndex: undefined,
    entireListIndexSetting: undefined,
    entireListTitleTemplate: undefined,
    entireListBodySetting: undefined,
    entireListBodyTemplate: undefined,
    eachItemIndex: undefined,
    eachItemTitleTemplate: undefined,
    eachItemBodySetting: undefined,
    eachItemBodyTemplate: undefined,
    fileAttachmentIndex: undefined,
    multiFolder: undefined,
    listId: undefined,
    discussionSettingEnum: undefined,
    containerPath: undefined,
    category: undefined,
}) {
    declare exception: string;
    declare domain: DomainDesign;
    declare name: string;
    declare domainKindName: string;
    declare description: string;
    declare lastIndexed: any;
    declare keyName: string;
    declare titleColumn: string;
    declare domainId: number;
    declare keyType: string;
    declare discussionSetting: number;
    declare allowDelete: true;
    declare allowUpload: true;
    declare allowExport: true;
    declare entireListIndex: true;
    declare entireListIndexSetting: number;
    declare entireListTitleTemplate: string;
    declare entireListBodySetting: number;
    declare entireListBodyTemplate: string;
    declare eachItemIndex: false;
    declare eachItemTitleTemplate: string;
    declare eachItemBodySetting: number;
    declare eachItemBodyTemplate: string;
    declare fileAttachmentIndex: false;
    declare multiFolder: false;
    declare listId: number;
    declare discussionSettingEnum: string;
    declare containerPath: string;
    declare category: string;

    static create(raw: any, defaultSettings = null): ListModel {
        if (defaultSettings) {
            const domain = DomainDesign.create(undefined);
            return new ListModel({ ...defaultSettings, domain });
        } else {
            let domain = DomainDesign.create(raw.domainDesign);

            // Issue39818: Set the key field of an existing list to be PKLocked.
            const fields = domain.fields;
            const pkField = fields.findIndex(i => i.isPrimaryKey);
            if (pkField > -1) {
                const pkFieldLocked = fields
                    .get(pkField)
                    .merge({ lockType: DOMAIN_FIELD_PRIMARY_KEY_LOCKED }) as DomainField;
                const updatedFields = fields.set(pkField, pkFieldLocked);
                domain = domain.set('fields', updatedFields) as DomainDesign;
            }

            return new ListModel({ ...raw.options, domain });
        }
    }

    getDomainKind(): string {
        if (this.category === PUBLIC_PICKLIST_CATEGORY || this.category === PRIVATE_PICKLIST_CATEGORY) {
            return PICKLIST;
        } else if (this.keyType === 'Varchar') {
            return VAR_LIST;
        } else if (this.keyType === 'Integer' || this.keyType === 'AutoIncrementInteger') {
            return INT_LIST;
        }

        return undefined;
    }

    hasValidKeyType(): boolean {
        return this.getDomainKind() !== undefined;
    }

    isNew(): boolean {
        return !this.listId;
    }

    isValid(): boolean {
        return this.hasValidProperties() && this.hasValidKeyType() && !this.domain.hasInvalidFields();
    }

    hasValidProperties(): boolean {
        return this.name !== undefined && this.name !== null && this.name.trim().length > 0;
    }

    getOptions(): Record<string, any> {
        const options = this.toJS();

        // Note: keyName is primarily set using <SetKeyFieldNamePanel/>'s onSelectionChange()
        // Setting keyName here covers the use-case where a user sets a Key Field, and then changes its name
        const keyField = this.domain.fields.find(field => field.isPrimaryKey);
        if (keyField && this.keyName !== keyField.name) {
            options.keyName = keyField.name;
        }

        delete options.exception;
        delete options.domain;
        return options;
    }
}
