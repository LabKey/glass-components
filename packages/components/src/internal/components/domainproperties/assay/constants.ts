export const PROPERTIES_HEADER_ID = 'assay-properties-hdr';
export const FORM_ID_PREFIX = 'assay-design-';

export const FORM_IDS = {
    ASSAY_NAME: FORM_ID_PREFIX + 'name',
    ASSAY_DESCRIPTION: FORM_ID_PREFIX + 'description',
    AUTO_LINK_TARGET: FORM_ID_PREFIX + 'autoCopyTargetContainerId',
    AUTO_LINK_CATEGORY: FORM_ID_PREFIX + 'autoLinkCategory',
    BACKGROUND_UPLOAD: FORM_ID_PREFIX + 'backgroundUpload',
    DETECTION_METHOD: FORM_ID_PREFIX + 'selectedDetectionMethod',
    EDITABLE_RUNS: FORM_ID_PREFIX + 'editableRuns',
    EDITABLE_RESULTS: FORM_ID_PREFIX + 'editableResults',
    METADATA_INPUT_FORMAT: FORM_ID_PREFIX + 'selectedMetadataInputFormat',
    PLATE_TEMPLATE: FORM_ID_PREFIX + 'selectedPlateTemplate',
    PROTOCOL_TRANSFORM_SCRIPTS: FORM_ID_PREFIX + 'protocolTransformScripts',
    QC_ENABLED: FORM_ID_PREFIX + 'qcEnabled',
    SAVE_SCRIPT_FILES: FORM_ID_PREFIX + 'saveScriptFiles',
    PLATE_METADATA: FORM_ID_PREFIX + 'plateMetadata',
    STATUS: FORM_ID_PREFIX + 'status',
};

export const BOOLEAN_FIELDS = [
    FORM_IDS.BACKGROUND_UPLOAD,
    FORM_IDS.EDITABLE_RUNS,
    FORM_IDS.EDITABLE_RESULTS,
    FORM_IDS.QC_ENABLED,
    FORM_IDS.SAVE_SCRIPT_FILES,
    FORM_IDS.PLATE_METADATA,
    FORM_IDS.STATUS,
];
