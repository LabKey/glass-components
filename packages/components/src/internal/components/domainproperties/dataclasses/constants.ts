export const DATACLASS_DOMAIN_SYSTEM_FIELDS = [
    {
        Name: 'Name',
        Label: 'Name',
        'Data Type': 'Text',
        Required: false,
        Description: 'Contains a short name for this data object.',
    },
    {
        Name: 'Description',
        Label: 'Description',
        'Data Type': 'Text',
        Required: false,
        Description: 'Contains a description for this data object',
    },
];
