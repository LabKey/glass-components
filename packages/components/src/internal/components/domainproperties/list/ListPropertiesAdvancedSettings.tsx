import React, { ChangeEvent, FC, memo, ReactNode, useCallback, useState } from 'react';
import classNames from 'classnames';

import { Modal } from '../../../Modal';

import { getSubmitButtonClass } from '../../../app/utils';

import { DomainFieldLabel } from '../DomainFieldLabel';

import { HelpLink } from '../../../util/helpLinks';

import { SelectInput } from '../../forms/input/SelectInput';

import { LabelHelpTip } from '../../base/LabelHelpTip';

import { DomainDesignerRadio } from '../DomainDesignerRadio';

import { AdvancedSettingsForm, EachItemSettings, EntireListSettings, ListModel } from './models';

export const DATA_INDEXING_TIP =
    'Not recommend for large lists with frequent updates, since updating any item will cause re-indexing of the entire list.';
export const DOCUMENT_TITLE_TIP =
    'Any text you want displayed and indexed as the search result title. There are no substitution parameters available for this title. Leave the document title blank to use the default title.';
export const CUSTOM_TEMPLATE_TIP = 'Example: ${Key} ${value}';

export const DISPLAY_TITLE_TIP = (
    <>
        Choose a field to identify this list when other lists or datasets have lookups into this list. When “Auto” is
        enabled, LabKey will select the title field for you by using:
        <ul>
            <li>The first non-lookup string column</li>
            <li>The primary key, if there are no string fields</li>
        </ul>
    </>
);
export const DISCUSSION_LINKS_TIP =
    'Optionally allow one or more discussion links to be shown on the details view of each list item.';
export const SEARCH_INDEXING_TIP = 'Controls how this list is indexed for search within LabKey Server.';

interface DisplayTitleProps {
    model: ListModel;
    onSelectChange: (name, formValue, selected) => void;
    titleColumn: string;
}

export const DisplayTitle: FC<DisplayTitleProps> = memo(({ model, onSelectChange, titleColumn }) => {
    const fields = model.domain.fields.filter(field => !field.isCalculatedField());
    const disabled = fields.size === 0;
    const placeholder = disabled ? 'No fields have been defined yet' : 'Auto';

    return (
        <div className="list__advanced-settings-modal__display-title">
            <SelectInput
                name="titleColumn"
                options={fields.toArray()}
                placeholder={placeholder}
                inputClass="" // This attr is necessary for proper styling
                valueKey="name"
                labelKey="name"
                disabled={disabled}
                onChange={onSelectChange}
                value={titleColumn}
            />
        </div>
    );
});

interface DiscussionLinksProps {
    discussionSetting: number;
    onRadioChange: (evt: any) => void;
}

// TODO: use RadioGroupInput instead
const DISCUSSION_RADIO_NAME = 'discussionSetting';
const DiscussionInputs: FC<DiscussionLinksProps> = memo(({ onRadioChange, discussionSetting }) => (
    <div className="form-group">
        <DomainDesignerRadio
            name={DISCUSSION_RADIO_NAME}
            value={0}
            checked={discussionSetting === 0}
            onChange={onRadioChange}
        >
            Disable discussions
        </DomainDesignerRadio>

        <DomainDesignerRadio
            name={DISCUSSION_RADIO_NAME}
            value={1}
            checked={discussionSetting === 1}
            onChange={onRadioChange}
        >
            Allow one discussion per item
        </DomainDesignerRadio>

        <DomainDesignerRadio
            name={DISCUSSION_RADIO_NAME}
            value={2}
            checked={discussionSetting === 2}
            onChange={onRadioChange}
        >
            Allow multiple discussions per item
        </DomainDesignerRadio>
    </div>
));

interface TitleIndexFieldProps {
    name: string;
    onInputChange: (evt: any) => void;
    titleTemplate: string;
}

const TitleIndexField: FC<TitleIndexFieldProps> = memo(({ name, titleTemplate, onInputChange }) => (
    <div>
        <DomainFieldLabel label="Document title" helpTipBody={DOCUMENT_TITLE_TIP} />
        <span>
            <input
                className="form-control list__advanced-settings-modal__text-field"
                id={name}
                type="text"
                placeholder="Use default"
                value={titleTemplate == null ? '' : titleTemplate}
                onChange={onInputChange}
            />
        </span>
    </div>
));

interface MetadataIndexFieldProps {
    indexSetting: number;
    name: string;
    onRadioChange: (evt: any) => void;
}

// Note: Currently the radio values go from high to low in order to correspond with the previous
// designer's order of radio options. Once the old designer is deprecated, we could change these to
// count from low to high.
// TODO: use RadioGroupInput instead
const MetadataIndexField: FC<MetadataIndexFieldProps> = memo(({ indexSetting, name, onRadioChange }) => (
    <div className="form-group">
        <DomainDesignerRadio name={name} value={2} checked={indexSetting === 2} onChange={onRadioChange}>
            Include both metadata and data
            <LabelHelpTip title="Warning">{DATA_INDEXING_TIP}</LabelHelpTip>
        </DomainDesignerRadio>
        <DomainDesignerRadio name={name} value={1} checked={indexSetting === 1} onChange={onRadioChange}>
            Include data only
            <LabelHelpTip title="Warning">{DATA_INDEXING_TIP}</LabelHelpTip>
        </DomainDesignerRadio>
        <DomainDesignerRadio name={name} value={0} checked={indexSetting === 0} onChange={onRadioChange}>
            Include metadata only (name and description of list and fields)
        </DomainDesignerRadio>
    </div>
));

interface IndexFieldProps {
    bodySetting: number;
    bodyTemplate: string;
    name: string;
    onInputChange: (evt: any) => void;
    onRadioChange: (evt: any) => void;
}

export const IndexField: FC<IndexFieldProps> = memo(props => {
    const { name, onRadioChange, bodySetting, bodyTemplate, onInputChange } = props;
    const id = name === 'entireListBodySetting' ? 'entireListBodyTemplate' : 'eachItemBodyTemplate';

    // TODO: Use RadioGroupInput instead
    return (
        <div>
            <div className="form-group">
                <DomainDesignerRadio name={name} value={0} checked={bodySetting === 0} onChange={onRadioChange}>
                    Index all non-PHI text fields
                </DomainDesignerRadio>
                <DomainDesignerRadio name={name} value={1} checked={bodySetting === 1} onChange={onRadioChange}>
                    Index all non-PHI fields (text, number, date, and boolean)
                </DomainDesignerRadio>
                <DomainDesignerRadio name={name} value={2} checked={bodySetting === 2} onChange={onRadioChange}>
                    Index using custom template
                    <LabelHelpTip>{CUSTOM_TEMPLATE_TIP}</LabelHelpTip>
                </DomainDesignerRadio>
            </div>

            {bodySetting === 2 && (
                <input
                    id={id}
                    type="text"
                    value={bodyTemplate}
                    onChange={onInputChange}
                    className="form-control list__advanced-settings-modal__custom-template-text-field"
                />
            )}
        </div>
    );
});

interface SingleDocumentIndexFieldsProps extends EntireListSettings {
    onInputChange: (evt: any) => void;
    onRadioChange: (evt: any) => void;
}

export const SingleDocumentIndexFields: FC<SingleDocumentIndexFieldsProps> = memo(props => {
    const {
        onRadioChange,
        onInputChange,
        entireListTitleTemplate,
        entireListIndexSetting,
        entireListBodySetting,
        entireListBodyTemplate,
    } = props;

    return (
        <div className="list__advanced-settings-modal__single-doc-fields">
            <TitleIndexField
                titleTemplate={entireListTitleTemplate}
                name="entireListTitleTemplate"
                onInputChange={onInputChange}
            />

            <MetadataIndexField
                name="entireListIndexSetting"
                onRadioChange={onRadioChange}
                indexSetting={entireListIndexSetting}
            />

            <IndexField
                name="entireListBodySetting"
                onRadioChange={onRadioChange}
                onInputChange={onInputChange}
                bodySetting={entireListBodySetting}
                bodyTemplate={entireListBodyTemplate}
            />
        </div>
    );
});

interface SeparateDocumentIndexFieldsProps extends EachItemSettings {
    onInputChange: (evt: any) => void;
    onRadioChange: (evt: any) => void;
}

export const SeparateDocumentIndexFields: FC<SeparateDocumentIndexFieldsProps> = memo(props => {
    const { onRadioChange, onInputChange, eachItemTitleTemplate, eachItemBodySetting, eachItemBodyTemplate } = props;

    return (
        <div className="list__advanced-settings-modal__single-doc-fields">
            <TitleIndexField
                titleTemplate={eachItemTitleTemplate}
                name="eachItemTitleTemplate"
                onInputChange={onInputChange}
            />

            <IndexField
                name="eachItemBodySetting"
                onRadioChange={onRadioChange}
                onInputChange={onInputChange}
                bodySetting={eachItemBodySetting}
                bodyTemplate={eachItemBodyTemplate}
            />
        </div>
    );
});

interface CollapsibleFieldsProps {
    checked: boolean;
    collapseFields: () => void;
    expandFields: (expandedSection: string) => void;
    expanded: boolean;
    identifier: string;
    onCheckboxChange: (name: string, checked: boolean) => void;
    title: string;
}

const CollapsibleFields: FC<CollapsibleFieldsProps> = memo(props => {
    const { checked, children, collapseFields, expandFields, expanded, identifier, onCheckboxChange, title } = props;
    const icon = classNames('fa', 'fa-lg', { 'fa-angle-down': expanded, 'fa-angle-right': !expanded });
    const expand = useCallback((): void => {
        // The logic in here could be moved up to the parent component.
        expandFields(expanded ? '' : identifier);
    }, [expandFields, expanded, identifier]);
    const onChange = useCallback(
        (event: ChangeEvent<HTMLInputElement>): void => {
            const checked_ = event.target.checked;
            onCheckboxChange(identifier, checked_);

            if (checked_ && !expanded) {
                expand();
            } else if (!checked_ && expanded) {
                collapseFields();
            }
        },
        [collapseFields, expand, expanded, identifier, onCheckboxChange]
    );

    return (
        <div className="list__advanced-settings-modal__collapsible-field">
            <span className={icon} onClick={expand} />

            <div className="form-group">
                <label>
                    <input checked={checked} onChange={onChange} type="checkbox" />

                    {title}
                </label>
            </div>

            {expanded && children}
        </div>
    );
});

interface SearchIndexingProps {
    eachItemIndex: boolean;
    eachItemIndexSettings: EachItemSettings;
    entireListIndex: boolean;
    entireListIndexSettings: EntireListSettings;
    fileAttachmentIndex: boolean;
    onCheckboxChange: (name, checked) => void;
    onInputChange: (evt: any) => void;
    onRadioChange: (evt: any) => void;
}

export const SearchIndexing: FC<SearchIndexingProps> = memo(props => {
    const {
        onRadioChange,
        onCheckboxChange,
        onInputChange,
        entireListIndex,
        entireListIndexSettings,
        eachItemIndex,
        eachItemIndexSettings,
        fileAttachmentIndex,
    } = props;
    const [expanded, setExpanded] = useState<string>('');
    const collapseFields = useCallback(() => setExpanded(''), []);
    const onIndexFileAttachmentsChange = useCallback(
        (event: ChangeEvent<HTMLInputElement>): void => {
            onCheckboxChange('fileAttachmentIndex', event.target.checked);
        },
        [onCheckboxChange]
    );

    return (
        <div>
            <CollapsibleFields
                expanded={expanded === 'entireListIndex'}
                title="Index entire list as a single document"
                identifier="entireListIndex"
                expandFields={setExpanded}
                collapseFields={collapseFields}
                checked={entireListIndex}
                onCheckboxChange={onCheckboxChange}
            >
                <SingleDocumentIndexFields
                    onRadioChange={onRadioChange}
                    onInputChange={onInputChange}
                    {...entireListIndexSettings}
                />
            </CollapsibleFields>

            <CollapsibleFields
                expanded={expanded === 'eachItemIndex'}
                title="Index each item as a separate document"
                identifier="eachItemIndex"
                expandFields={setExpanded}
                collapseFields={collapseFields}
                checked={eachItemIndex}
                onCheckboxChange={onCheckboxChange}
            >
                <SeparateDocumentIndexFields
                    onRadioChange={onRadioChange}
                    onInputChange={onInputChange}
                    {...eachItemIndexSettings}
                />
            </CollapsibleFields>

            <div className="list__advanced-settings-modal__index-checkbox form-group">
                <label>
                    <input checked={fileAttachmentIndex} onChange={onIndexFileAttachmentsChange} type="checkbox" />
                    Index file attachments
                </label>
            </div>
        </div>
    );
});

interface SettingsContainerProps {
    tipBody: string | JSX.Element;
    tipTitle?: string;
    title: string;
}

const SettingsContainer: FC<SettingsContainerProps> = memo(({ children, title, tipBody, tipTitle }) => (
    <div className="list__advanced-settings-modal__section-container">
        <div className="list__advanced-settings-modal__heading">
            <span className="list__bold-text"> {title} </span>
            <LabelHelpTip title={tipTitle || title}>{tipBody}</LabelHelpTip>
        </div>

        {children}
    </div>
));

interface AdvancedSettingsProps {
    applyAdvancedProperties: (advancedSettingsForm: AdvancedSettingsForm) => void;
    model: ListModel;
    title: string;
}

interface AdvancedSettingsState extends AdvancedSettingsForm {
    modalOpen?: boolean;
}

export class AdvancedSettings extends React.PureComponent<AdvancedSettingsProps, Partial<AdvancedSettingsState>> {
    constructor(props) {
        super(props);
        const initialState = this.getInitialState();

        this.state = {
            modalOpen: false,
            ...initialState,
        } as AdvancedSettingsState;
    }

    getInitialState = (): AdvancedSettingsState => {
        const model = this.props.model;

        return {
            titleColumn: model.titleColumn,
            discussionSetting: model.discussionSetting,
            fileAttachmentIndex: model.fileAttachmentIndex,

            // entire list
            entireListIndex: model.entireListIndex,
            // document title
            entireListTitleTemplate: model.entireListTitleTemplate,
            // metadata/data
            entireListIndexSetting: model.entireListIndexSetting,
            // index
            entireListBodySetting: model.entireListBodySetting,
            entireListBodyTemplate: model.entireListBodyTemplate,

            // each item
            eachItemIndex: model.eachItemIndex,
            // document title
            eachItemTitleTemplate: model.eachItemTitleTemplate,
            // index
            eachItemBodySetting: model.eachItemBodySetting,
            eachItemBodyTemplate: model.eachItemBodyTemplate,
        };
    };

    toggleModal = (isModalOpen: boolean): void => {
        this.setState({ modalOpen: isModalOpen });

        // If modal is re-opened, reset unsaved values
        if (isModalOpen) {
            this.setState(this.getInitialState());
        }
    };

    openModal = (): void => this.toggleModal(true);

    closeModal = (): void => this.toggleModal(false);

    onRadioChange = (e: ChangeEvent<HTMLInputElement>): void => {
        const name = e.currentTarget.name;
        const value = e.target.value;
        this.setState({ [name]: parseInt(value, 10) });
    };

    onCheckboxChange = (name, checked): void => {
        this.setState({ [name]: checked });
    };

    onInputChange = (e: ChangeEvent<HTMLInputElement>): void => {
        const id = e.target.id;
        const value = e.target.value;

        this.setState({ [id]: value });
    };

    onSelectChange = (name, formValue, selected): void => {
        const value = selected ? selected.name : '<AUTO>';
        this.setState({ [name]: value });
    };

    applyChanges = (): void => {
        // eslint-disable-next-line @typescript-eslint/no-unused-vars
        const { modalOpen, ...advancedSettingsForm } = this.state;
        this.props.applyAdvancedProperties(advancedSettingsForm as AdvancedSettingsForm);
        this.toggleModal(false);
    };

    render(): ReactNode {
        const {
            modalOpen,
            discussionSetting,
            fileAttachmentIndex,
            entireListIndex,
            entireListTitleTemplate,
            entireListIndexSetting,
            entireListBodySetting,
            eachItemIndex,
            eachItemTitleTemplate,
            eachItemBodySetting,
            titleColumn,
            entireListBodyTemplate,
            eachItemBodyTemplate,
        } = this.state;
        const { title, model } = this.props;
        const entireListIndexSettings = {
            entireListTitleTemplate,
            entireListIndexSetting,
            entireListBodySetting,
            entireListBodyTemplate,
        };
        const eachItemIndexSettings = {
            eachItemTitleTemplate,
            eachItemBodySetting,
            eachItemBodyTemplate,
        };
        const footer = (
            <>
                <button
                    className="domain-adv-footer domain-adv-cancel-btn btn btn-default"
                    onClick={this.closeModal}
                    type="button"
                >
                    Cancel
                </button>

                <HelpLink className="domain-adv-footer domain-adv-link" topic="createListOptions#advanced">
                    Get help with list settings
                </HelpLink>

                <button
                    className={`domain-adv-footer domain-adv-apply-btn btn btn-${getSubmitButtonClass()}`}
                    onClick={this.applyChanges}
                    type="button"
                >
                    Apply
                </button>
            </>
        );

        return (
            <>
                <button className="domain-field-float-right btn btn-default" onClick={this.openModal} type="button">
                    {title}
                </button>

                {modalOpen && (
                    <Modal footer={footer} onCancel={this.closeModal} title="Advanced List Settings">
                        <SettingsContainer title="Default Display Field" tipBody={DISPLAY_TITLE_TIP}>
                            <DisplayTitle
                                model={model}
                                onSelectChange={this.onSelectChange}
                                titleColumn={titleColumn}
                            />
                        </SettingsContainer>

                        <SettingsContainer title="Discussion Threads" tipBody={DISCUSSION_LINKS_TIP}>
                            <DiscussionInputs
                                onRadioChange={this.onRadioChange}
                                discussionSetting={discussionSetting}
                            />
                        </SettingsContainer>

                        <SettingsContainer title="Search Indexing Options" tipBody={SEARCH_INDEXING_TIP}>
                            <SearchIndexing
                                onRadioChange={this.onRadioChange}
                                onCheckboxChange={this.onCheckboxChange}
                                onInputChange={this.onInputChange}
                                entireListIndex={entireListIndex}
                                entireListIndexSettings={entireListIndexSettings}
                                eachItemIndex={eachItemIndex}
                                eachItemIndexSettings={eachItemIndexSettings}
                                fileAttachmentIndex={fileAttachmentIndex}
                            />
                        </SettingsContainer>
                    </Modal>
                )}
            </>
        );
    }
}
