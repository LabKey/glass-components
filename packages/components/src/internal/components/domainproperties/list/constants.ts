export const DATA_INDEXING_TIP =
    'Not recommend for large lists with frequent updates, since updating any item will cause re-indexing of the entire list.';
export const DOCUMENT_TITLE_TIP =
    'Any text you want displayed and indexed as the search result title (ex. ListName - ${Key} ${value}). Leave the document title blank to use the default title.';
export const CUSTOM_TEMPLATE_TIP = 'Example: ${Key} ${value}';

export const DISCUSSION_LINKS_TIP =
    'Optionally allow one or more discussion links to be shown on the details view of each list item.';
export const SEARCH_INDEXING_TIP = 'Controls how this list is indexed for search within LabKey Server.';
