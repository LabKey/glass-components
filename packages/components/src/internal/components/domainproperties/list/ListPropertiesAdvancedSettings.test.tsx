import React from 'react';
import { render } from '@testing-library/react';
import { userEvent } from '@testing-library/user-event';
import { List } from 'immutable';

import {
    SELECT_INPUT_DISABLED_SELECTOR,
    SELECT_INPUT_PLACEHOLDER_SELECTOR,
    SELECT_INPUT_SINGLE_VALUE_SELECTOR,
} from '../../forms/input/SelectInputTestUtils';
import { DEFAULT_LIST_SETTINGS } from '../../../../test/data/constants';
import getDomainDetailsJSON from '../../../../test/data/list-getDomainDetails.json';

import { DomainField } from '../models';

import { AdvancedSettings, DisplayTitle, SearchIndexing, IndexField } from './ListPropertiesAdvancedSettings';
import { ListModel } from './models';

const emptyNewModel = ListModel.create(null, DEFAULT_LIST_SETTINGS);
const populatedExistingModel = ListModel.create(getDomainDetailsJSON);

describe('AdvancedSettings', () => {
    test('default properties', () => {
        const advancedSettings = (
            <AdvancedSettings title="Advanced Settings" model={emptyNewModel} applyAdvancedProperties={jest.fn()} />
        );

        const container = render(advancedSettings).container;
        expect(container).toMatchSnapshot();
    });

    test('display title select dropdown with existing list', () => {
        render(<DisplayTitle model={populatedExistingModel} onSelectChange={jest.fn()} titleColumn="Name" />);

        expect(document.querySelector(SELECT_INPUT_SINGLE_VALUE_SELECTOR).textContent).toEqual('Name');
        expect(document.querySelectorAll(SELECT_INPUT_DISABLED_SELECTOR)).toHaveLength(0);
    });

    test('display title select dropdown with new list and no fields present', () => {
        render(<DisplayTitle model={emptyNewModel} onSelectChange={jest.fn()} titleColumn={null} />);

        expect(document.querySelector(SELECT_INPUT_PLACEHOLDER_SELECTOR).textContent).toEqual(
            'No fields have been defined yet'
        );
        expect(document.querySelectorAll(SELECT_INPUT_DISABLED_SELECTOR)).toHaveLength(1);
    });

    test('display title select dropdown with new list and some fields present', () => {
        const newModelWithOneField = emptyNewModel.setIn(
            ['domain', 'fields'],
            List.of(DomainField.create({ name: 'dummyField' }))
        ) as ListModel;

        render(<DisplayTitle model={newModelWithOneField} onSelectChange={jest.fn()} titleColumn={null} />);

        expect(document.querySelector(SELECT_INPUT_PLACEHOLDER_SELECTOR).textContent).toEqual('Auto');
        expect(document.querySelectorAll(SELECT_INPUT_DISABLED_SELECTOR)).toHaveLength(0);
    });

    test("either search indexing options 'index entire list' or 'index each item' may be open, but not both", async () => {
        render(
            <SearchIndexing
                onRadioChange={jest.fn()}
                onInputChange={jest.fn()}
                onCheckboxChange={jest.fn()}
                eachItemIndex={false}
                entireListIndexSettings={{
                    entireListBodySetting: undefined,
                    entireListBodyTemplate: undefined,
                    entireListIndexSetting: undefined,
                    entireListTitleTemplate: undefined,
                }}
                entireListIndex={false}
                eachItemIndexSettings={{
                    eachItemBodySetting: undefined,
                    eachItemBodyTemplate: undefined,
                    eachItemTitleTemplate: undefined,
                }}
                fileAttachmentIndex={false}
            />
        );

        await userEvent.click(document.querySelector('.fa-angle-right'));
        expect(document.querySelectorAll('input[id="entireListTitleTemplate"]')).toHaveLength(1);
        expect(document.querySelectorAll('input[id="eachItemTitleTemplate"]')).toHaveLength(0);

        await userEvent.click(document.querySelector('.fa-angle-right'));
        expect(document.querySelectorAll('input[id="entireListTitleTemplate"]')).toHaveLength(0);
        expect(document.querySelectorAll('input[id="eachItemTitleTemplate"]')).toHaveLength(1);
    });

    test("setting 'index using custom template' generates a text input field", () => {
        render(
            <IndexField
                name="entireListBodySetting"
                onRadioChange={jest.fn()}
                onInputChange={jest.fn()}
                bodySetting={2}
                bodyTemplate=""
            />
        );

        expect(
            document.querySelectorAll('input.list__advanced-settings-modal__custom-template-text-field')
        ).toHaveLength(1);
    });
});
