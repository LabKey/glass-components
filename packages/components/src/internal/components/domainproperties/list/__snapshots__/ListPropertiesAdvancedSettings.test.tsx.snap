// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AdvancedSettings default properties 1`] = `
<div>
  <button
    class="domain-field-float-right btn btn-default"
    type="button"
  >
    Advanced Settings
  </button>
</div>
`;
